crates/ra_ide/src/completion/complete_trait_impl.rs

```diff
@@ -0,0 +1,377 @@
+//! FIXME: write short doc here
+
+use crate::{
+    completion::{
+        CompletionContext, CompletionItem, CompletionItemKind, CompletionKind, Completions,
+    },
+    display::FunctionSignature,
+};
+
+use hir::{self, Docs, HasSource};
+use ra_syntax::{
+    ast::{self, edit},
+    AstNode, SyntaxKind, SyntaxNode, TextRange,
+};
+
+use ra_assists::utils::get_missing_impl_items;
+use ra_text_edit::TextEdit;
+
+pub(crate) fn complete_trait_impl(acc: &mut Completions, ctx: &CompletionContext) {
+    let trigger = ctx.token.ancestors().find(|p| match p.kind() {
```

Comment:
I think this might be overly eager in finding the trigger.

For example, I get the magic completion in this case, but I probably should not:

```rust
#[test]
    fn completes_only_on_top_level() {
        let completions = complete(
            r"
            trait Iterator {
                fn next(&mut self);
                fn for_each(self) {}
            }

            impl Iterator for () {
                fn next(&mut self) {
                    <|>
                }
            }
            ",
        );
        assert_debug_snapshot!(completions, @r###"[]"###);
    }
```

```diff
    1       │-[]
          1 │+[
          2 │+    CompletionItem {
          3 │+        label: "fn for_each(..)",
          4 │+        source_range: [210; 210),
          5 │+        delete: [169; 228),
          6 │+        insert: "fn for_each(self) {}",
          7 │+        kind: Method,
          8 │+        lookup: "for_each",
          9 │+    },
         10 │+]
```

Reply:
I also get this panic, which seems like it might be related?

![image](https://user-images.githubusercontent.com/1711539/74550811-49a9d180-4f52-11ea-9241-aef8839446d1.png)

Reply:
Ah, overlooked. Nice catch. I originally thought the search for the parent `impl_block` would catch this case, but after evaluating the syntax tree I now see why it failed. Anyways, in my search for a `trigger` node, I also include a `BLOCK_EXPR` node which jumps to adding no completions in the subsequent `match`. Additionally, I added a test for this case.

Let me know if you have other concerns.